File: 0310-minimum-height-trees/0310-minimum-height-trees.java
class Solution {
    public List<Integer> findMinHeightTrees(int n, int[][] edges) {
        if (n == 1) return Arrays.asList(0);
        if (n == 2) return Arrays.asList(0, 1);

        List<Set<Integer>> graph = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            graph.add(new HashSet<>());
        }

        for (int[] edge : edges) {
            int u = edge[0];
            int v = edge[1];
            graph.get(u).add(v);
            graph.get(v).add(u);
        }

        Queue<Integer> leaves = new LinkedList<>();
        for (int i = 0; i < n; i++) {
            if (graph.get(i).size() == 1) {
                leaves.add(i);
            }
        }

        int remainingNodes = n;

        while (remainingNodes > 2) {
            int levelSize = leaves.size();
            remainingNodes -= levelSize;

            for (int i = 0; i < levelSize; i++) {
                int leaf = leaves.poll();

                for (int neighbor : graph.get(leaf)) {
                    graph.get(neighbor).remove(leaf);

                    if (graph.get(neighbor).size() == 1) {
                        leaves.add(neighbor);
                    }
                }

                graph.get(leaf).clear();
            }
        }

        return new ArrayList<>(leaves);
    }
}

File: 0310-minimum-height-trees/0310-minimum-height-trees.kt
class Solution {
    fun findMinHeightTrees(n: Int, edges: Array<IntArray>): List<Int> {
        if (n <= 2) return (0 until n).toList()

        val graph = MutableList(n) { mutableSetOf<Int>() }

        for ((u, v) in edges) {
            graph[u].add(v)
            graph[v].add(u)
        }

        val leaves = ArrayDeque<Int>()
        for (i in graph.indices) {
            if (graph[i].size == 1) {
                leaves.add(i)
            }
        }

        var remainingNodes = n

        while (remainingNodes > 2) {
            val levelSize = leaves.size
            remainingNodes -= levelSize

            repeat(levelSize) {
                val leaf = leaves.removeFirst()

                for (neighbor in graph[leaf]) {
                    graph[neighbor].remove(leaf)

                    if (graph[neighbor].size == 1) {
                        leaves.add(neighbor)
                    }
                }

                graph[leaf].clear()
            }
        }

        return leaves.toList()
    }
}
